File: 프로그래머스/4/42897. 도둑질/도둑질.py
def solution(money):
    n = len(money)
    if n == 1: return money[0]

    # choose first -> ignore last
    dp_f = [0] * n
    dp_f[0] = money[0]
    dp_f[1] = max(money[0], money[1])
    
    for i in range(2, n-1):
        dp_f[i] = max(dp_f[i-2] + money[i], dp_f[i-1])
        
    # choose last -> ignore first
    dp_l = [0] * n
    dp_l[1] = money[1]
    for i in range(2, n):
        dp_l[i] = max(dp_l[i-2] + money[i], dp_l[i-1])
    
    
    return max(max(dp_f), max(dp_l))

File: 프로그래머스/4/42897. 도둑질/도둑질.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> money) {
    const int n = money.size();
    
    // 첫 집 털기 -> 마지막 집 안 털기
    vector<int> dp1(n, 0);
    dp1[0] = money[0];
    dp1[1] = dp1[0];
    for (int i=2; i<n-1; i++) {
        dp1[i] = max(dp1[i-2] + money[i], dp1[i-1]);
    }
        
    // 첫 집 안 털기
    vector<int> dp2(n, 0);
    dp2[0] = 0;
    dp2[1] = money[1];
    for (int i=2; i<n; i++) {
        dp2[i] = max(dp2[i-2] + money[i], dp2[i-1]);
    }
    
    return max(dp1[n-2], dp2[n-1]);
}